From 66f733b1695af6bbaaf5dd8c1bfe924abcdfb239 Mon Sep 17 00:00:00 2001
From: Julien Gilli <julien.gilli@joyent.com>
Date: Tue, 15 Aug 2017 18:08:38 -0700
Subject: [PATCH] VOLAPI-72 creating volume with unsupported size should result
 in error, not in volume being created with closest matching size

---
 lib/endpoints/volumes.js                      | 182 ++++++-----------
 lib/errors.js                                 |  34 +++-
 test/integration/invalid-parameters.test.js   |  11 +-
 ...ared-volumes-creation-invalid-size.test.js |  93 ---------
 .../nfs-shared-volumes-creation-size.test.js  | 192 ++++++++++++++++++
 ...s-shared-volumes-list-volume-sizes.test.js |   5 +-
 6 files changed, 297 insertions(+), 220 deletions(-)
 delete mode 100644 test/integration/nfs-shared-volumes-creation-invalid-size.test.js
 create mode 100644 test/integration/nfs-shared-volumes-creation-size.test.js

diff --git a/lib/endpoints/volumes.js b/lib/endpoints/volumes.js
index b65818c..769b29a 100644
--- a/lib/endpoints/volumes.js
+++ b/lib/endpoints/volumes.js
@@ -36,48 +36,8 @@ var APPLICATION_STATE;
 var NFS_SHARED_VOLUME_ZONE_USER_SCRIPT
     = fs.readFileSync(__dirname + '/../user-script.sh', 'utf8');
 
-var DEFAULT_NFS_SHARED_VOLUME_PACKAGE_SIZE_IN_MBS = 10 * units.MIBS_IN_GB;
-assert.number(DEFAULT_NFS_SHARED_VOLUME_PACKAGE_SIZE_IN_MBS,
-    'DEFAULT_NFS_SHARED_VOLUME_PACKAGE_SIZE_IN_MBS');
-
 var VOLUME_TICKETS_SCOPE = 'nfs_volume';
 
-function _selectBestPackage(requestedSize, packagesList, options, callback) {
-    assert.number(requestedSize, 'requestedSize');
-    assert.arrayOfObject(packagesList, 'packagesList');
-    assert.object(options, 'options');
-    assert.object(options.log, 'options.log');
-    assert.func(callback, 'callback');
-
-    var bestPackage;
-    var err;
-    var log = options.log;
-
-    packagesList.forEach(function updateBestPackage(candidatePackage) {
-        var candidateIsLargeEnough = candidatePackage.quota
-            >= requestedSize;
-        var candidateSmallerThanBest = bestPackage === undefined ||
-            candidatePackage.quota < bestPackage.quota;
-        var candidateFitsBetter = candidateIsLargeEnough &&
-            (bestPackage === undefined || candidateSmallerThanBest);
-
-        log.debug({package: candidatePackage}, 'considering package...');
-        if (candidateFitsBetter) {
-            log.debug({
-                package: candidatePackage,
-                oldBestFit: bestPackage
-            }, 'package is better fit than current best fit');
-            bestPackage = candidatePackage;
-        }
-    });
-
-    if (bestPackage === undefined) {
-        err = new Error('Could not find package');
-    }
-
-    callback(err, bestPackage);
-}
-
 //
 // opts is an object that must include at least:
 //
@@ -102,54 +62,6 @@ function getAllNfsSharedVolumesPackages(opts, callback) {
     });
 }
 
-function _getBestPackage(volumeParams, options, callback) {
-    assert.object(volumeParams, 'volumeParams');
-    assert.object(options, 'options');
-    assert.object(options.log, 'options.log');
-    assert.func(callback, 'callback');
-
-    var log = options.log;
-    var papiClient = options.papiClient;
-    assert.object(papiClient, 'papiClient');
-
-    var requestedSize = volumeParams.size;
-
-    if (requestedSize === undefined) {
-        requestedSize = DEFAULT_NFS_SHARED_VOLUME_PACKAGE_SIZE_IN_MBS;
-    }
-
-    var context = {};
-    vasync.pipeline({
-        funcs: [
-            function getNfsSharedVolumesPackages(ctx, next) {
-                getAllNfsSharedVolumesPackages({
-                    papiClient: papiClient
-                }, function onListDone(err, pkgs, count) {
-                    ctx.nfsSharedVolumesPkgs = pkgs;
-                    next(err);
-                });
-            },
-            function selectBestPackage(ctx, next) {
-                _selectBestPackage(requestedSize, ctx.nfsSharedVolumesPkgs,
-                    options, function onBestPackageSelected(err, bestPackage) {
-                        if (bestPackage !== undefined) {
-                            log.debug({package: bestPackage},
-                                'Best package found');
-                        } else {
-                            log.debug('Could not find best package');
-                        }
-
-                        ctx.bestPackage = bestPackage;
-                        next(err);
-                    });
-            }
-        ],
-        arg: context
-    }, function onBestPackageSelect(err) {
-        callback(err, context.bestPackage);
-    });
-}
-
 function _buildStorageVMPayload(volumeParams, storageVmUuid, imageUuid,
     billingPackage) {
     assert.object(volumeParams, 'volumeParams');
@@ -487,6 +399,63 @@ function createVolume(req, res, next) {
                 done(err);
             });
         },
+        function getStorageVmPackage(ctx, done) {
+            var getStorageVmPkgErr;
+            var requestedSize = req.params.size;
+
+            req.log.debug({volumeParams: volumeParams},
+                'Finding corresponding storage VM package');
+
+            getAllNfsSharedVolumesPackages({
+                papiClient: req._papiClient
+            }, function onListDone(getNfsVolPkgsErr, pkgs, count) {
+                var availableSizes;
+                var idx = 0;
+                var storageVmPkg;
+
+                if (getNfsVolPkgsErr || !pkgs || pkgs.length === 0) {
+                    done(new errors.InternalError(getNfsVolPkgsErr,
+                        'Could not get NFS volumes packages'));
+                    return;
+                }
+
+                if (requestedSize === undefined) {
+                    ctx.storageVmPkg = pkgs[0];
+                } else {
+                    availableSizes = pkgs.map(function getSizeFromPkg(pkg) {
+                        return pkg.quota;
+                    });
+
+                    for (idx = 0; idx < pkgs.length; ++idx) {
+                        storageVmPkg = pkgs[idx];
+                        if (storageVmPkg.quota === requestedSize) {
+                            req.log.debug({package: storageVmPkg},
+                                'Storage VM package found');
+                            ctx.storageVmPkg = storageVmPkg;
+                            break;
+                        }
+                    }
+                }
+
+                if (ctx.storageVmPkg === undefined) {
+                    getStorageVmPkgErr =
+                        new errors.VolumeSizeNotAvailableError(requestedSize,
+                            availableSizes.sort(function numSort(a, b) {
+                                if (a > b) {
+                                    return 1;
+                                } else if (a < b) {
+                                    return -1;
+                                }
+
+                                return 0;
+                            }));
+                } else {
+                    volumeParams.size = ctx.storageVmPkg.quota;
+                }
+
+                done(getStorageVmPkgErr);
+            });
+        },
         function loadVolumeReservations(ctx, done) {
             reservationModels.listVolumeReservations({
                 volumeName: volumeName,
@@ -574,33 +543,6 @@ function createVolume(req, res, next) {
 
             doLoadVolume();
         },
-        function getBestPackage(ctx, done) {
-            var options = {
-                log: req.log,
-                papiClient: req._papiClient
-            };
-
-            req.log.debug({volumeParams: volumeParams},
-                'Finding most suitable package');
-
-            _getBestPackage(volumeParams, options,
-                function onPackage(getPackageErr, bestPackage) {
-                    var err;
-
-                    if (!getPackageErr && bestPackage) {
-                        ctx.volumeObject.value.size = bestPackage.quota;
-                    }
-
-                    ctx.bestPackage = bestPackage;
-
-                    if (getPackageErr) {
-                        err = new errors.InternalError(getPackageErr,
-                            'Error when getting best package for volume');
-                    }
-
-                    done(err);
-                });
-        },
         /*
          * In order to avoid concurrent updates racing (recording the storage
          * VM's uuid in Moray and recording the storage VM's state changes in
@@ -619,13 +561,13 @@ function createVolume(req, res, next) {
         function createStorageVM(ctx, done) {
             assert.object(volumeParams, 'volumeParams');
             assert.uuid(CONFIG.nfsServerImageUuid, 'CONFIG.nfsServerImageUuid');
-            assert.object(ctx.bestPackage, 'ctx.bestPackage');
+            assert.object(ctx.storageVmPkg, 'ctx.storageVmPkg');
             assert.uuid(ctx.storageVmUuid, 'ctx.storageVmUuid');
 
             var storageVmPayload =
                 _buildStorageVMPayload(volumeParams, ctx.storageVmUuid,
                     CONFIG.nfsServerImageUuid,
-                    ctx.bestPackage);
+                    ctx.storageVmPkg);
             req.log.debug({vmPayload: storageVmPayload}, 'Creating storage VM');
 
             vmapiClient.createVm({
@@ -1011,7 +953,7 @@ function formatVolumeSizes(volumePkgs) {
     // return an array of:
     //
     //  {
-    //      "description": "100 GiB",
+    //      "type": "tritonnfs",
     //      "size": 102400
     //   },
     //
@@ -1022,8 +964,8 @@ function formatVolumeSizes(volumePkgs) {
         assert.number(pkg.quota, 'pkg.quota');
 
         return {
-            description: volumeSizeDescription(pkg),
-            size: pkg.quota
+            size: pkg.quota,
+            type: 'tritonnfs'
         };
     }).sort(function sortPkgs(a, b) {
         return (a.size - b.size);
diff --git a/lib/errors.js b/lib/errors.js
index 4a88e1d..58ab4b0 100644
--- a/lib/errors.js
+++ b/lib/errors.js
@@ -63,17 +63,47 @@ function ValidationError(causes) {
     restify.RestError.call(this, {
         restCode: 'ValidationError',
         statusCode: 409,
-        message: 'Validation error, causes: ' + causes,
+        message: 'Validation error, causes: ' + causes.join(', '),
         constructorOpt: ValidationError
     });
     this.name = 'ValidationError';
 }
 util.inherits(ValidationError, restify.RestError);
 
+function VolumeSizeNotAvailableError(size, availableSizes) {
+    assert.number(size, 'size');
+    assert.arrayOfNumber(availableSizes, 'availableSizes');
+
+    var message = 'Volume size ' + size + ' is not available. Available ' +
+        'sizes are: ' + availableSizes.join(', ');
+
+    restify.RestError.call(this, {
+        restCode: 'VolumeSizeNotAvailable',
+        statusCode: 409,
+        message: message,
+        constructorOpt: VolumeSizeNotAvailableError,
+        /*
+         * We specify a custom "body" property so that we can include the list
+         * of available volume sizes in its "availableSizes" property. Clients
+         * can use that extra information to output error messages that are
+         * relevant to their users without having to send a separate request to
+         * the ListVolumeSizes endpoint.
+         */
+        body: {
+            code: 'VolumeSizeNotAvailable',
+            message: message,
+            availableSizes: availableSizes
+        }
+    });
+    this.name = 'VolumeSizeNotAvailableError';
+}
+util.inherits(VolumeSizeNotAvailableError, restify.RestError);
+
 module.exports = {
     InternalError: restify.InternalError,
     ValidationError: ValidationError,
     VolumeAlreadyExistsError: VolumeAlreadyExistsError,
     VolumeInUseError: VolumeInUseError,
-    VolumeNotFoundError: VolumeNotFoundError
+    VolumeNotFoundError: VolumeNotFoundError,
+    VolumeSizeNotAvailableError: VolumeSizeNotAvailableError
 };
\ No newline at end of file
diff --git a/test/integration/invalid-parameters.test.js b/test/integration/invalid-parameters.test.js
index 0850d09..5c14174 100644
--- a/test/integration/invalid-parameters.test.js
+++ b/test/integration/invalid-parameters.test.js
@@ -21,10 +21,15 @@ var testVolumes = require('./lib/volumes');
 var CLIENTS;
 
 function confirmBogusParam(t, endpoint, err) {
+    var expectedErrMsg =
+        'Validation error, causes: Error: invalid parameter: bogus';
+
     t.ok(err, endpoint + ' w/ bad param should result in an error');
-    t.equal((err ? err.message : ''),
-        'Validation error, causes: Error: invalid parameter: bogus',
-        'expected error due to invalid parameter');
+
+    if (err) {
+        t.notEqual(err.message.indexOf(expectedErrMsg), -1,
+            'expected error due to invalid parameter');
+    }
 }
 
 test('setup', function (tt) {
diff --git a/test/integration/nfs-shared-volumes-creation-invalid-size.test.js b/test/integration/nfs-shared-volumes-creation-invalid-size.test.js
deleted file mode 100644
index b4aa17f..0000000
--- a/test/integration/nfs-shared-volumes-creation-invalid-size.test.js
+++ /dev/null
@@ -1,93 +0,0 @@
-/*
- * This Source Code Form is subject to the terms of the Mozilla Public
- * License, v. 2.0. If a copy of the MPL was not distributed with this
- * file, You can obtain one at http://mozilla.org/MPL/2.0/.
- */
-
-/*
- * Copyright (c) 2016, Joyent, Inc.
- */
-
-var assert = require('assert-plus');
-var test = require('tape');
-var vasync = require('vasync');
-
-var configLoader = require('../../lib/config-loader');
-
-var clientsSetup = require('./lib/clients-setup');
-var resources = require('./lib/resources');
-var testVolumes = require('./lib/volumes');
-
-var CONFIG = configLoader.loadConfigSync();
-
-var UFDS_ADMIN_UUID = CONFIG.ufdsAdminUuid;
-assert.string(UFDS_ADMIN_UUID, 'UFDS_ADMIN_UUID');
-
-var CLIENTS;
-var NFS_SHARED_VOLUMES_NAMES_PREFIX = 'nfs-shared-volumes';
-var NFS_SHARED_VOLUMES_TYPE_NAME = 'tritonnfs';
-
-var NETWORKS;
-
-test('setup', function (tt) {
-    tt.test('setup clients', function (t) {
-        clientsSetup.getApiClients(function onClientsSetup(err, clients) {
-            CLIENTS = clients;
-            t.end();
-        });
-    });
-
-    tt.test('setup networks', function (t) {
-        CLIENTS.napi.get('/networks',
-            function onListNetworks(err, networks) {
-                t.ifError(err);
-                t.ok(networks);
-                t.ok(Array.isArray(networks));
-                t.ok(networks.length > 1);
-                NETWORKS = networks;
-                t.end();
-            });
-    });
-});
-
-test('NFS shared volume creation with invalid size', function (tt) {
-    var volumeName =
-        resources.makeResourceName(NFS_SHARED_VOLUMES_NAMES_PREFIX);
-
-    tt.test('creating a nfs shared volume with invalid size should fail',
-        function (t) {
-            var INVALID_SIZES = ['invalid-size', '%$%#$%', '', 0, -42];
-
-            vasync.forEachParallel({
-                func: createVolumeWithInvalidSize,
-                inputs: INVALID_SIZES
-            }, function invalidSizesTested(err, results) {
-                t.end();
-            });
-
-            function createVolumeWithInvalidSize(invalidSize, callback) {
-                assert.func(callback, 'callback');
-
-                var expectedErrMsg = 'Validation error, causes: Error: ' +
-                    'Volume size: "' + invalidSize + '" is not a valid ' +
-                    'volume size';
-
-                var volumeParams = {
-                    name: volumeName,
-                    owner_uuid: UFDS_ADMIN_UUID,
-                    type: NFS_SHARED_VOLUMES_TYPE_NAME,
-                    networks: [NETWORKS[0].uuid],
-                    size: invalidSize
-                };
-
-                CLIENTS.volapi.createVolume(volumeParams,
-                    function onVolumeCreated(err, volume) {
-                        t.ok(err, 'volume creation should result in an error');
-                        t.ok(err.message.indexOf(expectedErrMsg) !== -1,
-                            'Error message should be: ' + expectedErrMsg);
-
-                        callback();
-                    });
-            }
-        });
-});
\ No newline at end of file
diff --git a/test/integration/nfs-shared-volumes-creation-size.test.js b/test/integration/nfs-shared-volumes-creation-size.test.js
new file mode 100644
index 0000000..d761902
--- /dev/null
+++ b/test/integration/nfs-shared-volumes-creation-size.test.js
@@ -0,0 +1,192 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+/*
+ * Copyright (c) 2017, Joyent, Inc.
+ */
+
+var assert = require('assert-plus');
+var test = require('tape');
+var vasync = require('vasync');
+
+var configLoader = require('../../lib/config-loader');
+
+var clientsSetup = require('./lib/clients-setup');
+var resources = require('./lib/resources');
+var testVolumes = require('./lib/volumes');
+
+var CONFIG = configLoader.loadConfigSync();
+
+var UFDS_ADMIN_UUID = CONFIG.ufdsAdminUuid;
+assert.string(UFDS_ADMIN_UUID, 'UFDS_ADMIN_UUID');
+
+var CLIENTS;
+var NFS_SHARED_VOLUMES_NAMES_PREFIX = 'nfs-shared-volumes';
+var NFS_SHARED_VOLUMES_TYPE_NAME = 'tritonnfs';
+var UFDS_ADMIN_FABRIC_NETWORK;
+
+test('setup', function (tt) {
+    tt.test('setup clients', function (t) {
+        clientsSetup.getApiClients(function onClientsSetup(err, clients) {
+            CLIENTS = clients;
+            t.end();
+        });
+    });
+
+    tt.test('setup networks', function (t) {
+        CLIENTS.napi.get('/networks?owner_uuid=' + UFDS_ADMIN_UUID +
+            '&fabric=true',
+            function onListNetworks(err, networks) {
+                t.ifError(err, 'listing fabric networks for owner ' +
+                    UFDS_ADMIN_UUID + ' should not error');
+                t.ok(networks, 'listing fabric networks for owner ' +
+                    UFDS_ADMIN_UUID + ' should result in a non-empty list of ' +
+                    'networks');
+                t.ok(Array.isArray(networks),
+                    'list of networks should be an array');
+                t.ok(networks.length === 1, 'owner ' + UFDS_ADMIN_UUID +
+                    ' should have only 1 fabric network');
+
+                UFDS_ADMIN_FABRIC_NETWORK = networks[0];
+
+                t.end();
+            });
+    });
+});
+
+test('NFS shared volume creation with invalid size', function (tt) {
+    var volumeName =
+        resources.makeResourceName(NFS_SHARED_VOLUMES_NAMES_PREFIX);
+
+    tt.test('creating a nfs shared volume with invalid size should fail',
+        function (t) {
+            var INVALID_SIZES = ['invalid-size', '%$%#$%', '', 0, -42];
+
+            vasync.forEachParallel({
+                func: createVolumeWithInvalidSize,
+                inputs: INVALID_SIZES
+            }, function invalidSizesTested(err, results) {
+                t.end();
+            });
+
+            function createVolumeWithInvalidSize(invalidSize, callback) {
+                assert.func(callback, 'callback');
+
+                var expectedErrMsg = 'Validation error, causes: Error: ' +
+                    'Volume size: "' + invalidSize + '" is not a valid ' +
+                    'volume size';
+
+                var volumeParams = {
+                    name: volumeName,
+                    owner_uuid: UFDS_ADMIN_UUID,
+                    type: NFS_SHARED_VOLUMES_TYPE_NAME,
+                    networks: [UFDS_ADMIN_FABRIC_NETWORK.uuid],
+                    size: invalidSize
+                };
+
+                CLIENTS.volapi.createVolume(volumeParams,
+                    function onVolumeCreated(err, volume) {
+                        t.ok(err, 'volume creation should result in an error');
+                        t.ok(err.message.indexOf(expectedErrMsg) !== -1,
+                            'Error message should be: ' + expectedErrMsg);
+
+                        callback();
+                    });
+            }
+        });
+});
+
+test('NFS shared volume creation with unavailable size', function (tt) {
+    var volumeName =
+        resources.makeResourceName(NFS_SHARED_VOLUMES_NAMES_PREFIX);
+
+    tt.test('creating nfs shared volume with unavailable size should fail',
+        function (t) {
+            vasync.pipeline({arg: {}, funcs: [
+                function getVolumeSizes(ctx, next) {
+                    CLIENTS.volapi.listVolumeSizes(
+                        function onListVolSizes(listVolSizesErr, volSizes) {
+                            t.ifError(listVolSizesErr,
+                                'listing volume sizes should not error');
+
+                            if (listVolSizesErr) {
+                                next(listVolSizesErr);
+                                return;
+                            }
+
+                            t.ok(volSizes && volSizes.length > 0,
+                                'there should be at least one ' +
+                                    'available volume size');
+
+                            if (!volSizes || volSizes.length === 0) {
+                                next(new Error('Could not find vol sizes'));
+                            } else {
+                                ctx.availableSizes =
+                                    volSizes.map(function getSize(volSize) {
+                                        return volSize.size;
+                                    });
+                                next();
+                            }
+                        });
+                },
+                function createVolWithUnavailableSize(ctx, next) {
+                    var expectedErrorCode;
+                    var expectedErrMsg;
+                    var expectedErrorName;
+                    var volumeSizeTooBig;
+
+                    assert.arrayOfNumber(ctx.availableSizes,
+                            'ctx.availableSizes');
+
+                    ctx.availableSizes.sort(function numSort(a, b) {
+                        if (a > b) {
+                            return 1;
+                        } else if (a < b) {
+                            return -1;
+                        }
+
+                        return 0;
+                    });
+
+                    volumeSizeTooBig =
+                        ctx.availableSizes[ctx.availableSizes.length - 1] + 1;
+
+                    expectedErrorCode = 'VolumeSizeNotAvailable';
+                    expectedErrMsg = 'Volume size ' + volumeSizeTooBig +
+                        ' is not available';
+                    expectedErrorName = 'VolumeSizeNotAvailableError';
+
+                    var volumeParams = {
+                        name: volumeName,
+                        owner_uuid: UFDS_ADMIN_UUID,
+                        type: NFS_SHARED_VOLUMES_TYPE_NAME,
+                        networks: [UFDS_ADMIN_FABRIC_NETWORK.uuid],
+                        size: volumeSizeTooBig
+                    };
+
+                    CLIENTS.volapi.createVolume(volumeParams,
+                        function onVolumeCreated(err, volume) {
+                            t.ok(err,
+                                'volume creation should result in an error');
+                            t.ok(err.message.indexOf(expectedErrMsg) !== -1,
+                                'Error message should be: ' + expectedErrMsg);
+                            t.equal(err.restCode, expectedErrorCode,
+                                'Error restCode should be ' +
+                                    expectedErrorCode);
+                            t.equal(err.body.code, expectedErrorCode,
+                                'Error\'s body\'s code should be ' +
+                                    expectedErrorCode);
+                            t.equal(err.name, expectedErrorName,
+                                'Error name should be ' +
+                                    expectedErrorName);
+                            next();
+                        });
+                }
+            ]}, function onTestDone() {
+                t.end();
+            });
+        });
+});
\ No newline at end of file
diff --git a/test/integration/nfs-shared-volumes-list-volume-sizes.test.js b/test/integration/nfs-shared-volumes-list-volume-sizes.test.js
index 4b11288..ee50232 100644
--- a/test/integration/nfs-shared-volumes-list-volume-sizes.test.js
+++ b/test/integration/nfs-shared-volumes-list-volume-sizes.test.js
@@ -45,8 +45,9 @@ test('listVolumeSizes', function (tt) {
                     'should have at least one volumeSize');
                 t.ok(volumeSizes[0].size, 'first volumeSize should have size,' +
                     ' got: ' + volumeSizes[0].size);
-                t.ok(volumeSizes[0].description, 'first volumeSize should ' +
-                    'have description, got: ' + volumeSizes[0].description);
+                t.equal(volumeSizes[0].type, 'tritonnfs',
+                    'first volumeSize should have type tritonnfs got: ' +
+                        volumeSizes[0].type);
 
                 // check that volume sizes are in ascending order
                 for (idx = 0; idx < volumeSizes.length; idx++) {
-- 
2.21.0

